src/linearAlgebra/MultiVector.t.cpp

```diff
@@ -632,5 +635,139 @@ namespace dftefe
+    {
+      return d_linAlgOpContext;
+    }
+
+    template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+    global_size_type
+    MultiVector<ValueType, memorySpace>::globalSize() const
+    {
+      return d_globalSize;
+    }
+
+    template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+    size_type
+    MultiVector<ValueType, memorySpace>::localSize() const
+    {
+      return d_localSize;
+    }
+
+    template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+    size_type
+    MultiVector<ValueType, memorySpace>::locallyOwnedSize() const
+    {
+      return d_locallyOwnedSize;
+    }
+
+    template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+    size_type
+    MultiVector<ValueType, memorySpace>::ghostSize() const
+    {
+      return d_ghostSize;
+    }
+
+    template <typename ValueType, dftefe::utils::MemorySpace memorySpace>
+    size_type
+    MultiVector<ValueType, memorySpace>::numVectors() const
+    {
+      return d_numVectors;
+    }
+
+    //
+    // Helper functions
+    //
+
+    template <typename ValueType1,
+              typename ValueType2,
+              utils::MemorySpace memorySpace>
+    void
+    add(blasLapack::scalar_type<ValueType1, ValueType2> a,
+        const MultiVector<ValueType1, memorySpace> &    u,
+        blasLapack::scalar_type<ValueType1, ValueType2> b,
+        const MultiVector<ValueType2, memorySpace> &    v,
+        MultiVector<blasLapack::scalar_type<ValueType1, ValueType2>,
+                    memorySpace> &                      w)
+    {
+      DFTEFE_AssertWithMsg(u.isCompatible(v),
+                           "u and v Vectors being added are not compatible.");
+      DFTEFE_AssertWithMsg(
+        u.isCompatible(w),
+        "Resultant MultiVector w = u + v is compatible with u.");
+      const size_type nv = u.numVectors();
+      blasLapack::axpby(u.localSize() * nv,
+                        a,
+                        u.data(),
+                        b,
+                        v.data(),
+                        w.data(),
+                        *(w.getLinAlgOpContext()));
+    }
+
+    template <typename ValueType1,
+              typename ValueType2,
+              utils::MemorySpace memorySpace>
+    void
+    dot(const MultiVector<ValueType1, memorySpace> &     u,
+        const MultiVector<ValueType2, memorySpace> &     v,
+        blasLapack::scalar_type<ValueType1, ValueType2> *dotProd,
+        const blasLapack::ScalarOp &opU /*= blasLapack::ScalarOp::Identity*/,
+        const blasLapack::ScalarOp &opV /*= blasLapack::ScalarOp::Identity*/)
+    {
+      DFTEFE_AssertWithMsg(
+        u.isCompatible(v),
+        "u and v MultiVectors used for dot product are not compatible.");
+      const size_type nv = u.numVectors();
+      utils::MemoryStorage<blasLapack::scalar_type<ValueType1, ValueType2>,
+                           memorySpace>
+        dotProdLocallyOwned(nv, 0.0);
```

Comment:
I think the usual use case would be that this memory space is already allocated for efficiency purposes, but we should still reset all values to zero.

Reply:
However we can address this in a subsequent PR after merging this one

Reply:
So the dotProdLocallyOwned is my sendbuf to MPIAllReduce and dotProd (pre-allocated by the user) is my recvbuf. One possibility to avoid another memory allocation is to use MPI_IN_PLACE in MPIAllReduce for the sendbuf. That would tell MPI that the send and recv buffers are the same. I'll have to make some modifications to the part of MPIWrapper for no-mpi, but it's an easy fix. What do you think?

Reply:
Yes it is better to always use preallocated buffers, as the dotProd calls could be inside a hot loop.